### ISSUE TITLE
modules/db_sqlite: error loading drouting 

### ISSUE THREAD

**Issue by jarrodb** [OpenSIPS/opensips#471]
##### opensips.cfg 
```
# ----- drouting -----
modparam("drouting", "db_url", "sqlite://var/db/opensips")
```
##### opensips.log
```
Apr 18 15:39:39 localhost opensips: INFO:drouting:dr_init: Dynamic-Routing - initializing
Apr 18 15:39:39 localhost opensips: ERROR:db_sqlite:db_sqlite_query: failed to prepare: (no such table: version)
Apr 18 15:39:39 localhost opensips: ERROR:db_sqlite:db_sqlite_get_query_rows: failed to prepare query
```
#####  opens incorrect sqlite db handler:
```
# ls -al /var/db
-rw-r--r--  1 root root      0 Apr 18 15:28 opensips?  <-- this is created
```

##### gdb
```
(gdb) break modules/db_sqlite/dbase.c:162
(gdb) print _h->url
$2 = {s = 0x7ffff7431bd0 "sqlite://var/db/opensips8", len = 24}
(gdb) print query_holder
$3 = {s = 0x852740 <sql_buf> "select table_version from version where table_name='dr_gateways's", len = 64}
167		if (ret!=SQLITE_OK)
(gdb) s
168			LM_ERR("failed to prepare: (%s)\n", sqlite3_errmsg(CON_CONNECTION(_h)));
(gdb) print ret
$4 = 1
```

##### Synopsis
Seems like the pointer arithmetic in val2str for type DB_STR (modules/db_sqlite/val.c:229) is off (the trailing s is from the previous sql_buf that dialog populates however the length is correct), but the other issue is its not accessing the proper db_url string that is set correctly.  It's weird that the other modules do not seem to have this issue.  

For example, dialog is the first to load, and the gdb printout of _h->url returns the same data for its query and everything else seems ok.  Perhaps there is an invalid or lacking CON_<X> macro call?

**Comment by bogdan-iancu** [comment]
@jarrodb , do you have the "version" table created in your DB ? as the drouting module is looking into it to check the versions of the dr related tables.

**Comment by jarrodb** [comment]
Absolutely, every other module finds its version properly and the SQLite query built (query_holder) runs successfully on my database. 

However, the real problem is its trying to access the wrong SQLite path as noted by the newly created file in the same path as my actual database when the SQLite library attempts to read.

**Comment by ionutrazvanionita** [comment]
@jarrodb https://gist.github.com/ionutrazvanionita/5f09ec2d96ddc60f6695 this patch fixes the bad '\0' issue.

**Comment by jarrodb** [comment]
Thanks for the help ionut, here is the gdb output from drouting.c as we discussed in IRC:

```
Breakpoint 1, dr_init () at drouting.c:1176
1176			if( head_start->db_url.s == 0 ) {
(gdb) n
1180			memcpy( head_start->db_url.s, db_url.s, db_url.len );
(gdb) print db_url.s
$1 = 0x8a2140 "sqlite://var/db/opensips"
(gdb) print head_start->db_url.s
$2 = 0x7fffeecf1e00 ""
(gdb) n
1182			init_head_w_extern_params();
(gdb) print head_start->db_url.s
$3 = 0x7fffeecf1e00 "sqlite://var/db/opensips\270\301\311\001"
(gdb) n
1184			head_start->partition.s = "Default";
(gdb) print head_start->db_url.s
$4 = 0x7fffeecf1e00 "sqlite://var/db/opensips\030"
(gdb) 
```

**Comment by jarrodb** [comment]
So digging into it more, I guess print is just looking for a terminating character \0, but still what is passed to sqlite3 for its call to openDatabase includes an extra byte on the end and that's whats causing the problem:

119274  in sqlite3.c                                                            
(gdb) s                                                                         
openDatabase (zFilename=0x7fffeecf30b8 "/var/db/opensips\b", ppDb=ppDb@entry=0x7fffffffe040, flags=flags@entry=6, zVfs=zVfs@entry=0x0) at sqlite3.c:121423

**Comment by nixon** [comment]
(I've never used opensips, and I haven't done any C coding in a decade.)

It seems to me that `new_db_id` takes a `str*` and assigns it to the `char* url` member of a `db_id` struct.  However, the `str*` relies on its `len` member, but the `url` member assumes a null-terminated string.  When the `db_id->url` is later passed to `sqlite3_open`, there's no guarantee that the string was null-terminated.

Is something like the following patch needed to ensure that `db_id` is always null-terminated for `sqlite3_open`?

```
diff --git a/db/db_id.c b/db/db_id.c
index 9efc0eb..c470324 100644
--- a/db/db_id.c
+++ b/db/db_id.c
@@ -240,7 +240,9 @@ struct db_id* new_db_id(const str* url)
        }

        /* store the original url */
-       ptr->url = url->s;
+       ptr->url = pkg_malloc(url->len+1);
+       strncpy(ptr->url, url->s, url->len);
+       ptr->url[url->len] = '\0';

        return ptr;

@@ -291,5 +293,6 @@ void free_db_id(struct db_id* id)
        if (id->password) pkg_free(id->password);
        if (id->host) pkg_free(id->host);
        if (id->database) pkg_free(id->database);
+       if (id->url) pkg_free(id->url);
        pkg_free(id);
 }
```

I butchered up `main.c` to create a simple testcase.  Without the above patch, the `assert` fails.  With the patch, the `id->url` returned from `new_db_id` is what I would expect.

```
diff --git a/main.c b/main.c
index c44ebf2..4ebe8a0 100644
--- a/main.c
+++ b/main.c
@@ -770,8 +770,29 @@ error:
  * \return don't return on sucess, -1 on error
  * \see main_loop
  */
+#include <assert.h>
 int main(int argc, char** argv)
 {
+       str s;
+       struct db_id* id;
+
+       /*init pkg mallocs (before parsing cfg but after parsing cmd line !)*/
+       if (init_pkg_mallocs()==-1)
+               goto error00;
+
+       // create a str object with a length of 23, but a longer string to
+       // simulate unallocated memory
+       //               11111111112222
+       //      12345678901234567890123
+       s.s  = "sqlite://tmp/0123456789ABCDEF";
+       s.len = 13          +10;
+
+       id = new_db_id(&s);
+
+       assert( strcmp(id->url, "sqlite://tmp/0123456789") == 0 );
+       exit(0);
+
+
```

**Comment by ionutrazvanionita** [comment]
fixed 0b6402c

**Comment by jarrodb** [comment]
Confirmed that @ionutrazvanionita fixed the issue.  Thanks!
